### PR TITLE
types/model: reintroduce Digest

### PR DESCRIPTION
**Pull request by bmizerany**

### CODE REVIEW

File: types/model/name.go [hunk 1]
```diff
@@ -308,3 +309,57 @@ func cutPromised(s, sep string) (before, after string, ok bool) {
...
+		return Digest{}, fmt.Errorf("unsupported digest type %q", typ)
+	}
+	d := Digest{
+		Type: DigestTypeSHA256,
+	}
+	n, err := hex.Decode(d.Sum[:], []byte(encSum))
+	if err != nil {
+		return Digest{}, err
+	}
+	if n != 32 {
```
Comment (BruceMacD) [inline types/model/name.go:350]:
```suggestion
        // sha256 is 256 bits, so expect 32 bytes
	if n != 32 {
```

File: types/model/name.go [hunk 2]
```diff
@@ -308,3 +309,57 @@ func cutPromised(s, sep string) (before, after string, ok bool) {
...
+}
+
+func (d Digest) String() string {
+	if d.Type == DigestTypeInvalid {
+		return ""
+	}
+	return fmt.Sprintf("sha256-%x", d.Sum)
+}
+
+func (d Digest) IsValid() bool {
```
Comment (BruceMacD) [inline types/model/name.go:363]: Would it be possible to get through ParseDigest without an error being thrown in the invalid case?

**Review by BruceMacD — APPROVED** [review]

File: types/model/name.go (hunk 1 above)
Comment (bmizerany) [reply types/model/name.go:350]: The hex package documents this.

File: types/model/name.go (hunk 2 above)
Comment (bmizerany) [reply types/model/name.go:363]: It's possible to construct a zero value Digest and pass it to something expected a valid digest

Comment (bmizerany) [reply types/model/name.go:363]: This is type is still a WIP. We can change later once we have a feel for its broader use in the codebase.

**Pull request merged**